File: backend/src/main/java/com/joying/chat/service/ChatPresenceService.java
package com.joying.chat.service;

import java.time.Instant;
import java.util.List;
import java.util.concurrent.TimeUnit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import com.joying.chat.broadcast.ChatBroadcaster;
import com.joying.chat.domain.ChatRoom;
import com.joying.chat.domain.ChatRoomMember;
import com.joying.chat.dto.PresenceUpdateEvent;
import com.joying.chat.repository.ChatRoomMemberRepository;

import lombok.RequiredArgsConstructor;

/**
 * 누가 지금 붙어 있는지.
 *
 * <p>상태를 Redis에 두는 이유는 서버가 여러 대일 때 다른 서버도 같은 답을 해야 하기
 * 때문이다. 만료를 걸어 두어 끊긴 것을 못 받아도 시간이 지나면 스스로 사라진다.
 */
@Service
@RequiredArgsConstructor
public class ChatPresenceService {

	private static final Logger log = LoggerFactory.getLogger(ChatPresenceService.class);

	private static final String PRESENCE_KEY_PREFIX = "chat:presence:";
	private static final String LAST_SEEN_KEY_PREFIX = "chat:last_seen:";

	/** 지금 보고 있는 방 */
	private static final String ACTIVE_ROOM_KEY_PREFIX = "chat:active_room:";

	private static final long ONLINE_TIMEOUT_SECONDS = 300L;

	/** 보고 있는 방은 더 짧게 잡는다. 화면을 떠난 것을 오래 붙들고 있으면 안 된다 */
	private static final long ACTIVE_ROOM_TIMEOUT_SECONDS = 60L;

	private final RedisTemplate<String, String> redisTemplate;
	private final ChatBroadcaster chatBroadcaster;
	private final ChatRoomMemberRepository chatRoomMemberRepository;

	public void setOnline(Long memberId) {
		String key = PRESENCE_KEY_PREFIX + memberId;
		boolean wasOnline = Boolean.TRUE.equals(redisTemplate.hasKey(key));

		redisTemplate.opsForValue().set(key, "online", ONLINE_TIMEOUT_SECONDS, TimeUnit.SECONDS);

		// 이미 온라인이었으면 알릴 것이 없다. 살아 있다는 신호가 주기적으로 오므로
		// 매번 알리면 같은 내용이 계속 나간다.
		if (!wasOnline) {
			broadcastPresenceUpdate(memberId, true, null);
		}
	}

	public void setOffline(Long memberId) {
		redisTemplate.delete(PRESENCE_KEY_PREFIX + memberId);

		Instant now = Instant.now();
		redisTemplate.opsForValue().set(
			LAST_SEEN_KEY_PREFIX + memberId, now.toString(), 7, TimeUnit.DAYS);

		broadcastPresenceUpdate(memberId, false, now);
	}

	public boolean isOnline(Long memberId) {
		return Boolean.TRUE.equals(redisTemplate.hasKey(PRESENCE_KEY_PREFIX + memberId));
	}

	public Instant getLastSeenAt(Long memberId) {
		String lastSeen = redisTemplate.opsForValue().get(LAST_SEEN_KEY_PREFIX + memberId);
		if (lastSeen == null) {
			return null;
		}
		try {
			return Instant.parse(lastSeen);
		} catch (Exception e) {
			return null;
		}
	}

	/**
	 * 살아 있다는 신호. 만료가 걸려 있으므로 주기적으로 다시 알려야 한다.
	 */
	public void heartbeat(Long memberId) {
		setOnline(memberId);
	}

	public void enterChatRoom(Long memberId, Long chatRoomId) {
		redisTemplate.opsForValue().set(ACTIVE_ROOM_KEY_PREFIX + memberId,
			String.valueOf(chatRoomId), ACTIVE_ROOM_TIMEOUT_SECONDS, TimeUnit.SECONDS);
	}

	public void leaveChatRoom(Long memberId) {
		redisTemplate.delete(ACTIVE_ROOM_KEY_PREFIX + memberId);
	}

	/**
	 * 이 사람이 지금 이 방을 보고 있는지.
	 *
	 * <p>보고 있으면 푸시를 보내지 않는다. 화면에 이미 떠 있는 것을 알림으로 또 알릴
	 * 이유가 없다.
	 */
	public boolean isViewingChatRoom(Long memberId, Long chatRoomId) {
		String activeChatRoomId = redisTemplate.opsForValue().get(ACTIVE_ROOM_KEY_PREFIX + memberId);
		return String.valueOf(chatRoomId).equals(activeChatRoomId);
	}

	public void refreshChatRoomActivity(Long memberId, Long chatRoomId) {
		enterChatRoom(memberId, chatRoomId);
	}

	/**
	 * 이 사람과 대화 중인 상대들에게 상태가 바뀐 것을 알린다.
	 *
	 * <p>참여 중인 방을 전부 훑으므로 방이 많으면 그만큼 조회가 나간다. 연결과 해제
	 * 때마다 돈다는 점도 같이 봐야 한다.
	 */
	private void broadcastPresenceUpdate(Long memberId, boolean isOnline, Instant lastSeenAt) {
		try {
			List<ChatRoomMember> chatRoomMembers = chatRoomMemberRepository.findByMemberId(memberId);

			for (ChatRoomMember chatRoomMember : chatRoomMembers) {
				ChatRoom chatRoom = chatRoomMember.getChatRoom();

				Long otherMemberId = memberId.equals(chatRoom.getBuyer().getMemberId())
					? chatRoom.getSeller().getMemberId()
					: chatRoom.getBuyer().getMemberId();

				if (otherMemberId == null) {
					continue;
				}

				PresenceUpdateEvent event = PresenceUpdateEvent.builder()
					.memberId(memberId)
					.isOnline(isOnline)
					.lastSeenAt(lastSeenAt)
					.build();

				chatBroadcaster.toUser(otherMemberId, "/queue/presence-update", event);
			}
		} catch (Exception e) {
			// 상태를 못 알려도 대화 자체는 된다. 여기서 터뜨려 연결 처리를 막지 않는다.
			log.warn("온라인 상태 알림 실패: memberId={}, error={}", memberId, e.getMessage());
		}
	}
}

File: backend/src/main/kotlin/com/joying/chat/service/ChatPresenceService.kt
package com.joying.chat.service

import com.joying.chat.dto.PresenceUpdateEvent
import com.joying.chat.repository.ChatRoomMemberRepository
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.concurrent.TimeUnit

/**
 * 채팅 온라인 상태 관리 Service
 *
 * Redis를 사용하여 사용자의 온라인 상태 추적
 *
 * 온라인 상태 설정 시점:
 * 1. WebSocket 연결 시 즉시 (WebSocketEventListener)
 * 2. Heartbeat 전송 시 (30초마다 TTL 갱신)
 *
 * 오프라인 상태 설정 시점:
 * 1. WebSocket 해제 시 즉시 (WebSocketEventListener)
 * 2. 5분간 heartbeat 없으면 자동 (Redis TTL 만료)
 */
@Service
class ChatPresenceService(
    private val redisTemplate: RedisTemplate<String, String>,
    private val messagingTemplate: SimpMessagingTemplate,
    private val chatRoomMemberRepository: ChatRoomMemberRepository
) {

    companion object {
        private const val PRESENCE_KEY_PREFIX = "chat:presence:"
        private const val LAST_SEEN_KEY_PREFIX = "chat:last_seen:"
        private const val ACTIVE_ROOM_KEY_PREFIX = "chat:active_room:"  // 사용자가 현재 보고 있는 채팅방
        private const val ONLINE_TIMEOUT_SECONDS = 300L // 5분
        private const val ACTIVE_ROOM_TIMEOUT_SECONDS = 60L // 1분 (더 짧게 설정)
    }

    /**
     * 사용자 온라인 상태 설정
     *
     * @param memberId 회원 ID
     */
    fun setOnline(memberId: Long) {
        val key = "$PRESENCE_KEY_PREFIX$memberId"
        val wasOnline = redisTemplate.hasKey(key) == true

        redisTemplate.opsForValue().set(key, "online", ONLINE_TIMEOUT_SECONDS, TimeUnit.SECONDS)

        // 오프라인에서 온라인으로 변경된 경우에만 이벤트 전송
        if (!wasOnline) {
            broadcastPresenceUpdate(memberId, isOnline = true)
        }
    }

    /**
     * 사용자 오프라인 상태 설정
     *
     * @param memberId 회원 ID
     */
    fun setOffline(memberId: Long) {
        val presenceKey = "$PRESENCE_KEY_PREFIX$memberId"
        val lastSeenKey = "$LAST_SEEN_KEY_PREFIX$memberId"

        // 온라인 상태 제거
        redisTemplate.delete(presenceKey)

        // 마지막 접속 시간 저장 (7일간 보관)
        val now = Instant.now()
        redisTemplate.opsForValue().set(
            lastSeenKey,
            now.toString(),
            7,
            TimeUnit.DAYS
        )

        // 오프라인 상태 변경 이벤트 전송
        broadcastPresenceUpdate(memberId, isOnline = false, lastSeenAt = now)
    }

    /**
     * 사용자 온라인 여부 확인
     *
     * @param memberId 회원 ID
     * @return 온라인 여부
     */
    fun isOnline(memberId: Long): Boolean {
        val key = "$PRESENCE_KEY_PREFIX$memberId"
        return redisTemplate.hasKey(key) == true
    }

    /**
     * 마지막 접속 시간 조회
     *
     * @param memberId 회원 ID
     * @return 마지막 접속 시간 (null이면 기록 없음)
     */
    fun getLastSeenAt(memberId: Long): Instant? {
        val key = "$LAST_SEEN_KEY_PREFIX$memberId"
        val lastSeen = redisTemplate.opsForValue().get(key) ?: return null

        return try {
            Instant.parse(lastSeen)
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Heartbeat (온라인 상태 갱신)
     *
     * WebSocket 연결 중 주기적으로 호출하여 온라인 상태 유지
     *
     * @param memberId 회원 ID
     */
    fun heartbeat(memberId: Long) {
        setOnline(memberId)
    }

    /**
     * 사용자가 특정 채팅방에 입장 (화면을 보고 있음)
     *
     * @param memberId 회원 ID
     * @param chatRoomId 채팅방 ID
     */
    fun enterChatRoom(memberId: Long, chatRoomId: Long) {
        val key = "$ACTIVE_ROOM_KEY_PREFIX$memberId"
        redisTemplate.opsForValue().set(key, chatRoomId.toString(), ACTIVE_ROOM_TIMEOUT_SECONDS, TimeUnit.SECONDS)
    }

    /**
     * 사용자가 채팅방에서 퇴장 (화면을 나감)
     *
     * @param memberId 회원 ID
     */
    fun leaveChatRoom(memberId: Long) {
        val key = "$ACTIVE_ROOM_KEY_PREFIX$memberId"
        redisTemplate.delete(key)
    }

    /**
     * 사용자가 특정 채팅방을 보고 있는지 확인
     *
     * @param memberId 회원 ID
     * @param chatRoomId 채팅방 ID
     * @return 해당 채팅방을 보고 있으면 true
     */
    fun isViewingChatRoom(memberId: Long, chatRoomId: Long): Boolean {
        val key = "$ACTIVE_ROOM_KEY_PREFIX$memberId"
        val activeChatRoomId = redisTemplate.opsForValue().get(key) ?: return false
        return activeChatRoomId == chatRoomId.toString()
    }

    /**
     * 채팅방 활성 상태 갱신 (Heartbeat용)
     *
     * 사용자가 채팅방 화면에 있으면 주기적으로 호출하여 TTL 갱신
     *
     * @param memberId 회원 ID
     * @param chatRoomId 채팅방 ID
     */
    fun refreshChatRoomActivity(memberId: Long, chatRoomId: Long) {
        enterChatRoom(memberId, chatRoomId)
    }

    /**
     * 온라인 상태 변경 이벤트를 해당 사용자와 채팅방을 공유하는 모든 사용자에게 전송
     *
     * @param memberId 상태가 변경된 회원 ID
     * @param isOnline 온라인 여부
     * @param lastSeenAt 마지막 접속 시간 (오프라인일 때만)
     */
    private fun broadcastPresenceUpdate(memberId: Long, isOnline: Boolean, lastSeenAt: Instant? = null) {
        try {
            // 해당 사용자가 참여 중인 모든 채팅방 조회
            val chatRoomMembers = chatRoomMemberRepository.findByMemberId(memberId)

            // 각 채팅방의 상대방에게 온라인 상태 변경 이벤트 전송
            chatRoomMembers.forEach { chatRoomMember ->
                val chatRoom = chatRoomMember.chatRoom

                // 상대방 찾기 (buyer 또는 seller)
                val otherMemberId = if (chatRoom.buyer.memberId == memberId) {
                    chatRoom.seller.memberId
                } else {
                    chatRoom.buyer.memberId
                }

                if (otherMemberId != null) {
                    val event = PresenceUpdateEvent(
                        memberId = memberId,
                        isOnline = isOnline,
                        lastSeenAt = lastSeenAt
                    )

                    // 상대방에게 WebSocket으로 전송
                    messagingTemplate.convertAndSendToUser(
                        otherMemberId.toString(),
                        "/queue/presence-update",
                        event
                    )
                }
            }
        } catch (e: Exception) {
            // 오류 발생 시 로깅만 하고 계속 진행 (온라인 상태 변경은 중요하지만 치명적이지 않음)
            println("[ChatPresenceService] 온라인 상태 브로드캐스트 실패: ${e.message}")
        }
    }
}
